### PR TITLE
Revisit APIs for accessing script variables

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@cston @amcasey @KevinH-MS

File: src/Scripting/CSharpTest/ScriptTests.cs [hunk 1]
```diff
@@ -234,14 +234,68 @@ public void TestCreateMethodDelegate()
...
+                ContinueWith("var b = 2u;").
+                ContinueWith("var a = 3m;").
+                ContinueWith("var x = a + b;").
+                ContinueWith("var X = Y;");
+
+            var state = await script.RunAsync(globals);
+
+            AssertEx.Equal(new[] { "a", "b", "a", "x", "X" }, state.Variables.Select(v => v.Name));
+            AssertEx.Equal(new object[] { '1', 2u, 3m, 5m, 20 }, state.Variables.Select(v => v.Value));
+            AssertEx.Equal(new Type[] { typeof(char), typeof(uint), typeof(decimal), typeof(decimal), typeof(int) }, state.Variables.Select(v => v.Type));
```
Comment (KevinH-MS) [inline src/Scripting/CSharpTest/ScriptTests.cs]: Consider inferred array type for all 3 above.

File: src/Scripting/Core/ScriptState.cs [hunk 2]
```diff
@@ -33,24 +40,83 @@ internal ScriptState(ScriptExecutionState executionState, Script script)
...
         /// </summary>
-        public ScriptVariables Variables
+        public ImmutableArray<ScriptVariable> Variables
         {
             get
             {
                 if (_lazyVariables == null)
                 {
-                    Interlocked.CompareExchange(ref _lazyVariables, new ScriptVariables(ExecutionState), null);
+                    ImmutableInterlocked.InterlockedInitialize(ref _lazyVariables, CreateVariables());
```
Comment (KevinH-MS) [inline src/Scripting/Core/ScriptState.cs]: Neat!  I didn't know that method existed.

File: src/Scripting/Core/ScriptState.cs [hunk 3]
```diff
@@ -33,24 +40,83 @@ internal ScriptState(ScriptExecutionState executionState, Script script)
...
+            }
+
+            return result.ToImmutable();
+        }
+
+        private IReadOnlyDictionary<string, int> GetVariableMap()
+        {
+            if (_lazyVariableMap == null)
+            {
+                var map = new Dictionary<string, int>(Script.Compiler.IdentifierComparer);
```
Comment (KevinH-MS) [inline src/Scripting/Core/ScriptState.cs]: Can we use PooledDictionary here and store as ImmutableDictionary?

File: src/Scripting/Core/ScriptVariable.cs [hunk 4]
```diff
@@ -10,62 +10,58 @@ namespace Microsoft.CodeAnalysis.Scripting
...
+
+                _field.SetValue(_instance, value);
             }
         }
 
-        private string GetDebuggerDisplay()
-        {
-            return string.Format("{0}: {1}", this.Name, this.Value?.ToString() ?? "<null>");
-        }
+        private string GetDebuggerDisplay() => $"{Name}: {Value ?? "<null>"}";
```
Comment (KevinH-MS) [inline src/Scripting/Core/ScriptVariable.cs]: In the EE, we don't bracket the "null".

**Comment by KevinH-MS** [comment]
LGTM :+1:

**Pull request merged**